File: docs/reference/python-engines/integration/orchestrator.py
"""
Engine Orchestrator - Multi-Engine Workflow System

Coordinates multiple divination engines to create comprehensive readings
and consciousness field analysis.
"""

import asyncio
from typing import Dict, List, Any, Optional, Union
from datetime import datetime
import logging
from concurrent.futures import ThreadPoolExecutor, as_completed

try:
    from ..base.engine_interface import BaseEngine
    from ..base.data_models import BaseEngineInput, BaseEngineOutput, EngineError
    from .. import get_engine, list_engines
except ImportError:
    # Fallback for direct execution
    import sys
    import os
    sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
    from base.engine_interface import BaseEngine
    from base.data_models import BaseEngineInput, BaseEngineOutput, EngineError

    def get_engine(name):
        return None

    def list_engines():
        return ['numerology', 'biorhythm', 'human_design', 'vimshottari', 'gene_keys', 'tarot', 'iching']


class EngineOrchestrator:
    """
    Orchestrates multiple engines for complex divination workflows
    """
    
    def __init__(self, max_workers: int = 4):
        """Initialize the orchestrator with thread pool for parallel execution"""
        self.max_workers = max_workers
        self.logger = logging.getLogger(__name__)
        self.active_engines = {}
        self.workflow_cache = {}
        
    def load_engine(self, engine_name: str, config: Optional[Dict] = None) -> BaseEngine:
        """Load and cache an engine instance"""
        if engine_name not in self.active_engines:
            try:
                engine_class = get_engine(engine_name)
                self.active_engines[engine_name] = engine_class(config)
                self.logger.info(f"Loaded engine: {engine_name}")
            except Exception as e:
                raise EngineError(f"Failed to load engine {engine_name}: {str(e)}")
        
        return self.active_engines[engine_name]
    
    def run_single_engine(self, engine_name: str, input_data: BaseEngineInput, 
                         config: Optional[Dict] = None) -> BaseEngineOutput:
        """Run a single engine with input data"""
        engine = self.load_engine(engine_name, config)
        return engine.process(input_data)
    
    def run_parallel_engines(self, engine_configs: List[Dict]) -> Dict[str, BaseEngineOutput]:
        """
        Run multiple engines in parallel
        
        Args:
            engine_configs: List of dicts with 'name', 'input', and optional 'config'
        """
        results = {}
        
        with ThreadPoolExecutor(max_workers=self.max_workers) as executor:
            # Submit all engine tasks
            future_to_engine = {}
            for config in engine_configs:
                engine_name = config['name']
                input_data = config['input']
                engine_config = config.get('config')
                
                future = executor.submit(
                    self.run_single_engine, 
                    engine_name, 
                    input_data, 
                    engine_config
                )
                future_to_engine[future] = engine_name
            
            # Collect results as they complete
            for future in as_completed(future_to_engine):
                engine_name = future_to_engine[future]
                try:
                    result = future.result()
                    results[engine_name] = result
                    self.logger.info(f"Completed engine: {engine_name}")
                except Exception as e:
                    self.logger.error(f"Engine {engine_name} failed: {str(e)}")
                    results[engine_name] = EngineError(f"Engine failed: {str(e)}")
        
        return results
    
    def run_sequential_engines(self, engine_configs: List[Dict]) -> Dict[str, BaseEngineOutput]:
        """
        Run engines sequentially, allowing later engines to use earlier results
        """
        results = {}
        
        for config in engine_configs:
            engine_name = config['name']
            input_data = config['input']
            engine_config = config.get('config')
            
            # Allow input to reference previous results
            if hasattr(input_data, 'previous_results'):
                input_data.previous_results = results
            
            try:
                result = self.run_single_engine(engine_name, input_data, engine_config)
                results[engine_name] = result
                self.logger.info(f"Completed sequential engine: {engine_name}")
            except Exception as e:
                self.logger.error(f"Sequential engine {engine_name} failed: {str(e)}")
                results[engine_name] = EngineError(f"Engine failed: {str(e)}")
                # Continue with other engines even if one fails
        
        return results
    
    def create_comprehensive_reading(self, birth_data: Dict, 
                                   engines: Optional[List[str]] = None) -> Dict[str, Any]:
        """
        Create a comprehensive reading using multiple engines
        
        Args:
            birth_data: Birth information (date, time, location, name)
            engines: List of engine names to use (default: all available)
        """
        if engines is None:
            engines = ['numerology', 'biorhythm', 'human_design', 'vimshottari', 
                      'gene_keys', 'tarot', 'iching']
        
        # Prepare engine configurations
        engine_configs = []
        
        for engine_name in engines:
            if engine_name in ['numerology', 'biorhythm']:
                # These engines need basic birth data
                input_data = self._prepare_basic_input(birth_data, engine_name)
            elif engine_name in ['human_design', 'vimshottari']:
                # These need astronomical calculations
                input_data = self._prepare_astro_input(birth_data, engine_name)
            elif engine_name in ['gene_keys']:
                # Gene Keys uses Human Design foundation
                input_data = self._prepare_gene_keys_input(birth_data)
            elif engine_name in ['tarot', 'iching']:
                # Divination engines need intention/question
                input_data = self._prepare_divination_input(birth_data, engine_name)
            else:
                continue
            
            engine_configs.append({
                'name': engine_name,
                'input': input_data
            })
        
        # Run engines in parallel for independent calculations
        results = self.run_parallel_engines(engine_configs)
        
        # Add metadata
        reading = {
            'timestamp': datetime.now().isoformat(),
            'birth_data': birth_data,
            'engines_used': engines,
            'results': results,
            'synthesis': None  # Will be filled by ResultSynthesizer
        }
        
        return reading
    
    def _prepare_basic_input(self, birth_data: Dict, engine_name: str) -> BaseEngineInput:
        """Prepare input for basic engines (numerology, biorhythm)"""
        # This would be implemented based on each engine's input model
        # For now, return a placeholder
        return BaseEngineInput()
    
    def _prepare_astro_input(self, birth_data: Dict, engine_name: str) -> BaseEngineInput:
        """Prepare input for astronomical engines"""
        # This would be implemented based on each engine's input model
        return BaseEngineInput()
    
    def _prepare_gene_keys_input(self, birth_data: Dict) -> BaseEngineInput:
        """Prepare input for Gene Keys engine"""
        return BaseEngineInput()
    
    def _prepare_divination_input(self, birth_data: Dict, engine_name: str) -> BaseEngineInput:
        """Prepare input for divination engines"""
        return BaseEngineInput()
    
    def get_available_engines(self) -> List[str]:
        """Get list of available engines"""
        return list_engines()
    
    def clear_cache(self):
        """Clear workflow cache"""
        self.workflow_cache.clear()
        self.logger.info("Workflow cache cleared")

File: src/integration/orchestrator.ts
/**
 * Engine Orchestrator - Multi-Engine Workflow System
 * 
 * Coordinates multiple divination engines to create comprehensive readings
 * and consciousness field analysis.
 */

import { BaseEngineInput, BaseEngineOutput, CalculationResult } from '../engines/core/types';
import { getEngine, listEngines } from '../engines/index';
import type { EngineName } from '../types/engines';

export interface EngineConfig {
  name: EngineName;
  input: BaseEngineInput;
  config?: Record<string, any>;
}

export interface ComprehensiveReading {
  timestamp: string;
  birthData?: Record<string, any>;
  enginesUsed: EngineName[];
  results: Record<string, CalculationResult<BaseEngineOutput>>;
  synthesis?: any; // Will be filled by ResultSynthesizer
  processingTime: number;
}

export interface WorkflowResult {
  workflowName: string;
  timestamp: string;
  results: Record<string, CalculationResult<BaseEngineOutput>>;
  synthesis: any;
  recommendations: string[];
  processingTime: number;
}

export class EngineOrchestrator {
  private maxWorkers: number;
  private activeEngines: Map<EngineName, any> = new Map();
  private workflowCache: Map<string, WorkflowResult> = new Map();

  constructor(maxWorkers: number = 4) {
    this.maxWorkers = maxWorkers;
  }

  /**
   * Load and cache an engine instance
   */
  private loadEngine(engineName: EngineName): any {
    if (!this.activeEngines.has(engineName)) {
      try {
        const engine = getEngine(engineName);
        this.activeEngines.set(engineName, engine);
        console.log(`Loaded engine: ${engineName}`);
      } catch (error) {
        throw new Error(`Failed to load engine ${engineName}: ${error}`);
      }
    }
    return this.activeEngines.get(engineName);
  }

  /**
   * Run a single engine with input data
   */
  async runSingleEngine(
    engineName: EngineName, 
    inputData: BaseEngineInput, 
    config?: Record<string, any>
  ): Promise<CalculationResult<BaseEngineOutput>> {
    const engine = this.loadEngine(engineName);
    return engine.calculate(inputData);
  }

  /**
   * Run multiple engines in parallel
   */
  async runParallelEngines(engineConfigs: EngineConfig[]): Promise<Record<string, CalculationResult<BaseEngineOutput>>> {
    const results: Record<string, CalculationResult<BaseEngineOutput>> = {};
    
    // Create promises for all engine calculations
    const enginePromises = engineConfigs.map(async (config) => {
      try {
        const result = await this.runSingleEngine(config.name, config.input, config.config);
        return { name: config.name, result };
      } catch (error) {
        console.error(`Engine ${config.name} failed:`, error);
        return { 
          name: config.name, 
          result: {
            success: false,
            error: {
              code: 'ENGINE_ERROR',
              message: error instanceof Error ? error.message : 'Unknown error',
              context: { input: config.input },
              suggestions: ['Check input parameters'],
              timestamp: new Date().toISOString()
            },
            processingTime: 0,
            timestamp: new Date().toISOString()
          } as CalculationResult<BaseEngineOutput>
        };
      }
    });

    // Wait for all engines to complete
    const engineResults = await Promise.all(enginePromises);
    
    // Organize results by engine name
    engineResults.forEach(({ name, result }) => {
      results[name] = result;
    });

    return results;
  }

  /**
   * Run engines sequentially, allowing later engines to use earlier results
   */
  async runSequentialEngines(engineConfigs: EngineConfig[]): Promise<Record<string, CalculationResult<BaseEngineOutput>>> {
    const results: Record<string, CalculationResult<BaseEngineOutput>> = {};
    
    for (const config of engineConfigs) {
      try {
        // Allow input to reference previous results
        const enhancedInput = {
          ...config.input,
          previousResults: results
        };
        
        const result = await this.runSingleEngine(config.name, enhancedInput, config.config);
        results[config.name] = result;
        console.log(`Completed sequential engine: ${config.name}`);
      } catch (error) {
        console.error(`Sequential engine ${config.name} failed:`, error);
        results[config.name] = {
          success: false,
          error: {
            code: 'ENGINE_ERROR',
            message: error instanceof Error ? error.message : 'Unknown error',
            context: { input: config.input },
            suggestions: ['Check input parameters'],
            timestamp: new Date().toISOString()
          },
          processingTime: 0,
          timestamp: new Date().toISOString()
        };
        // Continue with other engines even if one fails
      }
    }
    
    return results;
  }

  /**
   * Create a comprehensive reading using multiple engines
   */
  async createComprehensiveReading(
    birthData: Record<string, any>, 
    engines?: EngineName[]
  ): Promise<ComprehensiveReading> {
    const startTime = Date.now();
    
    if (!engines) {
      engines = ['numerology', 'biorhythm', 'human_design', 'vimshottari', 
                 'gene_keys', 'tarot', 'iching', 'enneagram', 'sacred_geometry'];
    }

    // Prepare engine configurations
    const engineConfigs: EngineConfig[] = [];
    
    for (const engineName of engines) {
      let inputData: BaseEngineInput;
      
      if (['numerology', 'biorhythm'].includes(engineName)) {
        // Basic birth data engines
        inputData = this.prepareBasicInput(birthData, engineName);
      } else if (['human_design', 'vimshottari'].includes(engineName)) {
        // Astronomical calculation engines
        inputData = this.prepareAstroInput(birthData, engineName);
      } else if (['gene_keys'].includes(engineName)) {
        // Gene Keys uses Human Design foundation
        inputData = this.prepareGeneKeysInput(birthData);
      } else if (['tarot', 'iching'].includes(engineName)) {
        // Divination engines
        inputData = this.prepareDivinationInput(birthData, engineName);
      } else if (['enneagram'].includes(engineName)) {
        // Personality assessment engines
        inputData = this.preparePersonalityInput(birthData, engineName);
      } else if (['sacred_geometry', 'sigil_forge'].includes(engineName)) {
        // Symbolic/geometric engines
        inputData = this.prepareSymbolicInput(birthData, engineName);
      } else {
        continue;
      }
      
      engineConfigs.push({
        name: engineName as EngineName,
        input: inputData
      });
    }

    // Run engines in parallel for independent calculations
    const results = await this.runParallelEngines(engineConfigs);
    
    const reading: ComprehensiveReading = {
      timestamp: new Date().toISOString(),
      birthData,
      enginesUsed: engines as EngineName[],
      results,
      synthesis: null, // Will be filled by ResultSynthesizer
      processingTime: Date.now() - startTime
    };
    
    return reading;
  }

  /**
   * Prepare input for basic engines (numerology, biorhythm)
   */
  private prepareBasicInput(birthData: Record<string, any>, engineName: string): BaseEngineInput {
    return {
      birthDate: birthData.birthDate || birthData.date,
      fullName: birthData.fullName || birthData.name,
      birthTime: birthData.birthTime || birthData.time,
      birthLocation: birthData.birthLocation || birthData.location
    };
  }

  /**
   * Prepare input for astronomical engines
   */
  private prepareAstroInput(birthData: Record<string, any>, engineName: string): BaseEngineInput {
    return {
      birthDate: birthData.birthDate || birthData.date,
      birthTime: birthData.birthTime || birthData.time || '12:00',
      birthLocation: birthData.birthLocation || birthData.location || 'New York, NY',
      fullName: birthData.fullName || birthData.name
    };
  }

  /**
   * Prepare input for Gene Keys engine
   */
  private prepareGeneKeysInput(birthData: Record<string, any>): BaseEngineInput {
    return {
      birthDate: birthData.birthDate || birthData.date,
      birthTime: birthData.birthTime || birthData.time || '12:00',
      birthLocation: birthData.birthLocation || birthData.location || 'New York, NY',
      focusSequence: 'activation'
    };
  }

  /**
   * Prepare input for divination engines
   */
  private prepareDivinationInput(birthData: Record<string, any>, engineName: string): BaseEngineInput {
    const question = `Life guidance for ${birthData.fullName || birthData.name || 'this person'}`;
    
    if (engineName === 'tarot') {
      return {
        question,
        spreadType: 'three_card',
        includeReversed: true
      };
    } else if (engineName === 'iching') {
      return {
        question,
        method: 'coins'
      };
    }
    
    return { question };
  }

  /**
   * Prepare input for personality engines
   */
  private preparePersonalityInput(birthData: Record<string, any>, engineName: string): BaseEngineInput {
    if (engineName === 'enneagram') {
      return {
        identificationMethod: 'intuitive',
        behavioralDescription: `Person seeking comprehensive personality analysis`,
        includeWings: true,
        includeArrows: true,
        includeInstincts: true
      };
    }
    
    return {};
  }

  /**
   * Prepare input for symbolic engines
   */
  private prepareSymbolicInput(birthData: Record<string, any>, engineName: string): BaseEngineInput {
    if (engineName === 'sacred_geometry') {
      return {
        birthDate: birthData.birthDate || birthData.date,
        patternType: 'auto',
        includePersonalization: true
      };
    } else if (engineName === 'sigil_forge') {
      return {
        intention: `Manifestation support for ${birthData.fullName || birthData.name}`,
        generationMethod: 'personal',
        includePersonalization: true
      };
    }
    
    return {};
  }

  /**
   * Get list of available engines
   */
  getAvailableEngines(): EngineName[] {
    return listEngines();
  }

  /**
   * Clear engine cache
   */
  clearCache(): void {
    this.activeEngines.clear();
    this.workflowCache.clear();
  }
}

export const orchestrator = new EngineOrchestrator();
export default orchestrator;
